### ISSUE TITLE
official-api branch release soon?

### ISSUE THREAD

**Issue by shadoi** [CenturyLinkLabs/fleet-api#7]
Since 0.9.0 is out and it's possible to enable the API via cloud-init is this ready for use?

**Comment by bdehamer** [comment]
@shadoi I've sorta been waiting for Fleet 0.9.0 to land on the stable branch of CoreOS. Shortly after that happens we'll get the official API stuff merged to master.

**Comment by shadoi** [comment]
It made it into beta!  I'd be happy to test and provide patches to help it along, I've got a couple apps that really need to move away from fleetctl via ssh.

**Comment by bdehamer** [comment]
@shadoi I'm going to start working on this tomorrow. Please feel free to do some testing and let me know how it goes.

**Comment by shadoi** [comment]
I gave it a quick shot a few days ago.  I ran into issues with the path getting generated in Request for Unit and/or Job, Machine was working.

**Comment by bdehamer** [comment]
@shadoi I just pushed a change to the official API branch which should fix the path issue

**Comment by shadoi** [comment]
OK, now I'm up against my original issue, systemd unit files with multiple ExecStartPre lines.  I found IniParse which handles these duplicate options, but now I have to figure out how to properly format them for Fleet::ServiceDefinition to be happy.  Maybe I'll send a patch with some helpers to convert from an IniParse::Document (unless you have some better tools somewhere for this?)

**Comment by shadoi** [comment]
Actually it looks like ServiceDefinition expects a hash like: `{sections: [{ name: value }]}` which won't handle duplicate keys either... :/

**Comment by bdehamer** [comment]
I've got a fix for this, but it's on the master branch (haven't been good at keeping the two in sync). I'll work to get all the fixes on master ported back to this branch.

**Comment by shadoi** [comment]
Great, I faked it out to get past it and now I get this:

```ruby
Fleet::BadRequest: unable to decode body: json: cannot unmarshal object into Go value of type []*schema.UnitOption
	~/.bundler/ruby/2.2.0/fleet-api-8f6eb6b9aa9f/lib/fleet/middleware/response/raise_error.rb:20:in `on_complete'
	/usr/local/lib/ruby/gems/2.2.0/gems/faraday-0.8.9/lib/faraday/response.rb:9:in `block in call'
	/usr/local/lib/ruby/gems/2.2.0/gems/faraday-0.8.9/lib/faraday/response.rb:63:in `on_complete'
	/usr/local/lib/ruby/gems/2.2.0/gems/faraday-0.8.9/lib/faraday/response.rb:8:in `call'
	/usr/local/lib/ruby/gems/2.2.0/gems/faraday_middleware-0.9.0/lib/faraday_middleware/response_middleware.rb:30:in `call'
	/usr/local/lib/ruby/gems/2.2.0/gems/faraday_middleware-0.9.0/lib/faraday_middleware/request/encode_json.rb:23:in `call'
	/usr/local/lib/ruby/gems/2.2.0/gems/faraday-0.8.9/lib/faraday/connection.rb:253:in `run_request'
	/usr/local/lib/ruby/gems/2.2.0/gems/faraday-0.8.9/lib/faraday/connection.rb:118:in `put'
	/Users/blake_barnett/.bundler/ruby/2.2.0/fleet-api-8f6eb6b9aa9f/lib/fleet/request.rb:17:in `request'
	/Users/blake_barnett/.bundler/ruby/2.2.0/fleet-api-8f6eb6b9aa9f/lib/fleet/request.rb:10:in `block (2 levels) in <module:Request>'
	~/.bundler/ruby/2.2.0/fleet-api-8f6eb6b9aa9f/lib/fleet/client/unit.rb:16:in `create_unit'
	~/.bundler/ruby/2.2.0/fleet-api-8f6eb6b9aa9f/lib/fleet/client.rb:34:in `load'
```

I forced it to use to_json on the body in Fleet::Request and that seems to have solved it.  I'll keep hacking.  :)

**Comment by shadoi** [comment]
Next blocker: can't start named services via templates.  I can load up a unit for `service-name@.service`, but I can't just start an instance of it with `fleet.start("service-name@1.service")`, etc.

Maybe this isn't supported via the REST API yet?  I'll take a closer look in a bit.

**Comment by bdehamer** [comment]
Multiple values for any command now supported with the format documented in the readme

```
If you need mutiple values for a single statement (like multiple `ExecStart` instructions) you can use an array of strings:

	{
	  'Unit' => {
	    'Description' => 'Useless infinite loop'
	  },
	  'Service' => {
	    'ExecStart' => ["/bin/bash -c \"while true; do sleep 1; done\"", "some other command"]
	  }
	}
```

**Comment by shadoi** [comment]
Cool, I'll try that out.  For now I just made this patch, and then pass in properly formatted options:
```ruby
diff --git a/lib/fleet/service_definition.rb b/lib/fleet/service_definition.rb
index f1d4b1a..5aee18f 100644
--- a/lib/fleet/service_definition.rb
+++ b/lib/fleet/service_definition.rb
@@ -6,10 +6,16 @@ module Fleet
     end

     def to_unit
-      {
-        'desiredState' => 'loaded',
-        'options' => options
-      }
+      h = { 'machineID' => '',
+        'currentState' => 'inactive',
+        'desiredState' => 'loaded' }
+
+      if (@service_def.is_a?(Hash) && @service_def.has_key?('options'))
+        h['options'] = @service_def['options']
+      else
+        h['options'] = options
+      end
+      h
     end

     private
```

**Comment by bdehamer** [comment]
I just pushed a couple more fixes (mainly bringing this branch up-to-date with features that had been added on master recently). My next task is to look at the template stuff -- I don't believe that I ever had that working on any of these branches so it may take a bit of investigation to figure it out. I'll keep you posted.

**Comment by shadoi** [comment]
Yeah, I filed this bug because I can’t see a way to do it with the REST API at all right now: https://github.com/coreos/fleet/issues/1129 <https://github.com/coreos/fleet/issues/1129>

**Comment by bdehamer** [comment]
From my investigation, it looks like the template stuff is really a feature of the `fleetctl` client and not the API itself. If you do something like `fleetctl start foo@1.service` it results in API traffic that looks like this:

    GET /fleet/v1/units/foo@1.service // which results in a 404
    GET /fleet/v1/units/foo@service 
    PUT /fleet/v1/units/foo@1.service
    GET /fleet/v1/units/foo@1.service // 200, this time

Basically, `fleetctl` will pull down the template and then turn around and submit it as a new unit with the specified name.

Given that this is really an implementation detail for the Fleet client and not part of the API itself, I can't decide if it's something that should be in the fleet-api gem. A user certainly could implement the same behavior using the operations that are exposed through the fleet-api gem.

**Comment by bdehamer** [comment]
In case you're interested, here's the full `ngrep` trace when trying to start a template unit:

```
root@8baf9502ea25:/# fleetctl --driver=API start f@9.service

T 10.1.0.2:45283 -> 10.1.42.1:49153 [AP]
GET http://10.1.42.1:49153/fleet/v1/units/f%409.service?alt=json HTTP/1.1.
Host: 10.1.42.1:49153.
User-Agent: google-api-go-client/0.5.
Accept-Encoding: gzip.
.

T 10.1.42.1:49153 -> 10.1.0.2:45283 [AP]
HTTP/1.1 404 Not Found.
Content-Type: application/json.
Server: fleetd/0.9.0.
Date: Mon, 16 Feb 2015 18:35:02 GMT.
Content-Length: 54.
.
{"error":{"code":404,"message":"unit does not exist"}}


T 10.1.0.2:45283 -> 10.1.42.1:49153 [AP]
GET http://10.1.42.1:49153/fleet/v1/units/f%40.service?alt=json HTTP/1.1.
Host: 10.1.42.1:49153.
User-Agent: google-api-go-client/0.5.
Accept-Encoding: gzip.
.

T 10.1.42.1:49153 -> 10.1.0.2:45283 [AP]
HTTP/1.1 200 OK.
Content-Type: application/json.
Server: fleetd/0.9.0.
Date: Mon, 16 Feb 2015 18:35:02 GMT.
Content-Length: 263.
.
{"currentState":"inactive","desiredState":"inactive","name":"f@.service","options":[{"name":"Description","section":"Unit","value":"infinit loop param"},{"name":"ExecStart","section":"Service","value":"/bin/bash -c \"while true; do echo 'HI'; sleep %i; done\""}]}


T 10.1.0.2:45283 -> 10.1.42.1:49153 [AP]
PUT http://10.1.42.1:49153/fleet/v1/units/f%409.service?alt=json HTTP/1.1.
Host: 10.1.42.1:49153.
User-Agent: google-api-go-client/0.5.
Content-Length: 213.
Content-Type: application/json.
Accept-Encoding: gzip.
.
{"name":"f@9.service","options":[{"name":"Description","section":"Unit","value":"infinit loop param"},{"name":"ExecStart","section":"Service","value":"/bin/bash -c \"while true; do echo 'HI'; sleep %i; done\""}]}

T 10.1.42.1:49153 -> 10.1.0.2:45283 [AP]
HTTP/1.1 201 Created.
Server: fleetd/0.9.0.
Date: Mon, 16 Feb 2015 18:35:02 GMT.
Content-Length: 0.
Content-Type: text/plain; charset=utf-8.
.


T 10.1.0.2:45283 -> 10.1.42.1:49153 [AP]
GET http://10.1.42.1:49153/fleet/v1/units/f%409.service?alt=json HTTP/1.1.
Host: 10.1.42.1:49153.
User-Agent: google-api-go-client/0.5.
Accept-Encoding: gzip.
.

T 10.1.42.1:49153 -> 10.1.0.2:45283 [AP]
HTTP/1.1 200 OK.
Content-Type: application/json.
Server: fleetd/0.9.0.
Date: Mon, 16 Feb 2015 18:35:02 GMT.
Content-Length: 264.
.
{"currentState":"inactive","desiredState":"inactive","name":"f@9.service","options":[{"name":"Description","section":"Unit","value":"infinit loop param"},{"name":"ExecStart","section":"Service","value":"/bin/bash -c \"while true; do echo 'HI'; sleep %i; done\""}]}


T 10.1.0.2:45283 -> 10.1.42.1:49153 [AP]
PUT http://10.1.42.1:49153/fleet/v1/units/f%409.service?alt=json HTTP/1.1.
Host: 10.1.42.1:49153.
User-Agent: google-api-go-client/0.5.
Content-Length: 49.
Content-Type: application/json.
Accept-Encoding: gzip.
.
{"desiredState":"launched","name":"f@9.service"}

T 10.1.42.1:49153 -> 10.1.0.2:45283 [AP]
HTTP/1.1 204 No Content.
Server: fleetd/0.9.0.
Date: Mon, 16 Feb 2015 18:35:02 GMT.
.
```

**Comment by shadoi** [comment]
Yeah, I’m going to go down this path for now. Maybe I’ll write it as a patch?  It certainly would be convenient to have it work by default.

**Comment by dserfez** [comment]
Kudos for the fleet-api gem, and it would really be nice to have templates working straight out from the fleet-api gem.

**Comment by bdehamer** [comment]
Version 1.0.0 of the fleet-api gem has been released which now makes use of the official Fleet API (no more direct interaction with etcd!).

Please open a new issue for any problems discovered in the 1.0.0 code.
